oracle/oracle.go

```diff
@@ -24,252 +27,131 @@ type Oracle interface {
+	//
+	// cfg is the oracle configuration.
+	cfg config.OracleConfig
+	// marketMap is the market map that the oracle is using.
+	marketMap mmtypes.MarketMap
+	// writeTo is a path to write the market map to.
+	writeTo string
+
+	// -------------------Provider Constructor Fields-------------------//
+	//
+	// priceAPIFactory factory is a factory function that creates price API query handlers.
+	priceAPIFactory types.PriceAPIQueryHandlerFactory
+	// priceWSFactory is a factory function that creates price websocket query handlers.
+	priceWSFactory types.PriceWebSocketQueryHandlerFactory
+	// marketMapperFactory is a factory function that creates market map providers.
+	marketMapperFactory mmclienttypes.MarketMapFactory
+
+	// -------------------Metrics Fields-------------------//
+	//
+	// wsMetrics is the web socket metrics.
+	wsMetrics wsmetrics.WebSocketMetrics
+	// apiMetrics is the API metrics.
+	apiMetrics apimetrics.APIMetrics
+	// providerMetrics is the provider metrics.
+	providerMetrics providermetrics.ProviderMetrics
+	// metrics are the base metrics of the oracle.
 	metrics oraclemetrics.Metrics
+}
 
-	// updateInterval is the interval at which the oracle will fetch prices from
-	// each provider.
-	updateInterval time.Duration
-
-	// maxCacheAge is the longest amount of time a price will stay in our cache
-	maxCacheAge time.Duration
+// ProviderState is the state of a provider. This includes the provider implementation,
+// the provider specific market map, and whether the provider is enabled.
+type ProviderState struct {
+	// Provider is the price provider implementation.
+	Provider *types.PriceProvider
+	// Cfg is the provider configuration.
+	//
+	// TODO: Deprecate this once we have synchronous configuration updates.
+	Cfg config.ProviderConfig
 }
 
-// New returns a new instance of an Oracle. The oracle inputs providers that are
-// responsible for fetching prices for a given set of currency pairs (base, quote). The oracle
-// will fetch new prices concurrently every oracleTicker interval. In the case where
-// the oracle fails to fetch prices from a given provider, it will continue to fetch prices
-// from the remaining providers. The oracle currently assumes that each provider aggregates prices
-// using TWAPs, TVWAPs, etc. When determining final prices, the oracle will utilize the aggregateFn
-// to compute the final price for each currency pair. By default, the oracle will compute the median
-// price across all providers.
-func New(opts ...Option) (*OracleImpl, error) {
-	o := &OracleImpl{
-		closer:         ssync.NewCloser(),
-		logger:         zap.NewNop(),
-		metrics:        oraclemetrics.NewNopMetrics(),
-		updateInterval: 1 * time.Second,
-		maxCacheAge:    time.Minute, // default max cache age is 1 minute
+// New returns a new Oracle.
+func New(
+	cfg config.OracleConfig,
+	aggregator PriceAggregator,
+	opts ...Option,
+) (Oracle, error) {
+	if err := cfg.ValidateBasic(); err != nil {
+		return nil, err
 	}
```

Comment:
mega nit: would add nil check on the aggregator.

Reply:
oops. fixed here [add read lock to GetLastSyncTime and nil check aggregator](https://github.com/skip-mev/slinky/pull/503/commits/8cb65abfcb2f0d9f70bfcc91acb30b11a7416198)